build_tools/buildkite/github/post_benchmarks_as_pr_comment.py

```diff
@@ -0,0 +1,152 @@
+
+This script is meant to be used by Buildkite for automation. It requires the
+following environment to be set:
+
+- BUILDKITE_BUILD_URL: the link to the current buildkite build.
+- BUILDKITE_COMMIT: the pull request HEAD commit.
+- BUILDKITE_PULL_REQUEST: the current pull request number.
+- GITHUB_TOKEN: personal access token to authenticate against GitHub API.
+
+This script uses pip package "markdown_strings".
+
+Example usage:
+  # Export necessary environment variables:
+  export ...
+  # Then run the script:
+  python3 post_benchmarks_as_pr_comment.py <benchmark-json-file>..
+  #   where each <benchmark-json-file> is expected to be of format:
+  #   {"commit": <commit-sha>, "benchmarks": {<name>: <value>, ...}}
+"""
+
+import argparse
+import json
+import os
+import requests
+import markdown_strings as md
+
+GITHUB_IREE_API_PREFIX = "https://api.github.com/repos/google/iree"
+
+
+def get_required_env_var(var):
+  """Gets the value for a required environment variable."""
+  value = os.getenv(var, None)
+  if value is None:
+    raise RuntimeError(f'Missing environment variable "{var}"')
+  return value
+
+
+def aggregate_all_benchmarks(benchmark_files):
+  """Aggregates all benchmarks in the given files.
+
+  Args:
+  - benchmark_files: A list of JSON files, each with the format:
+    {"commit": <commit-sha>, "benchmarks": {<name>: <value>, ...}}
+
+  Returns:
+  - A list of benchmark (name, value) tuples.
+  """
+  pr_commit = get_required_env_var("BUILDKITE_COMMIT")
+  all_benchmarks = []
+
+  for benchmark_file in benchmark_files:
+    with open(benchmark_file) as f:
+      benchmarks = json.load(f)
+      if benchmarks["commit"] != pr_commit:
+        raise ValueError("Inconsistent pull request commit")
+      for name, value in benchmarks["benchmarks"].items():
+        if name in all_benchmarks:
+          raise ValueError(f"Duplicated benchmarks: {name}")
+        all_benchmarks.append((name, value))
+
+  return sorted(all_benchmarks)
+
+
+def get_benchmark_result_markdown(benchmark_files):
+  """Gets markdown summary of all benchmarks in the given files."""
+  all_benchmarks = aggregate_all_benchmarks(benchmark_files)
+  names = [k for k, _ in all_benchmarks]
+  values = [v for _, v in all_benchmarks]
+  names.insert(0, "Benchmark Name")
+  values.insert(0, "Time (ms)")
+
+  build_url = get_required_env_var("BUILDKITE_BUILD_URL")
+  pr_commit = get_required_env_var("BUILDKITE_COMMIT")
+
+  commit = f"@ commit {pr_commit}"
+  header = md.header("Benchmark results", 3)
+  benchmark_table = md.table([names, values])
+  link = "See more details on " + md.link("Buildkite", build_url)
+
+  return "\n\n".join([header, commit, benchmark_table, link])
+
+
+def comment_on_pr(content):
+  """Posts the given content as comments to the current pull request."""
+  pr_number = get_required_env_var("BUILDKITE_PULL_REQUEST")
+  if pr_number == "false":
+    raise ValueError("Not a pull request")
```

Comment:
Link to https://buildkite.com/docs/pipelines/environment-variables#bk-env-vars-buildkite-pull-request. This was confusing to me :-D

```suggestion
  # Buildkite sets this to "false" if not running on a PR:
  # https://buildkite.com/docs/pipelines/environment-variables#bk-env-vars-buildkite-pull-request
  if pr_number == "false":
    raise ValueError("Not a pull request")
```